src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala

```diff
@@ -21,30 +37,43 @@ object RichDenseMatrixDouble {
     val data = new Array[Double](rows * cols)
     in.readDoubles(data)
 
-    new DenseMatrix[Double](rows, cols, data,
+    new BDM[Double](rows, cols, data,
       offset = 0, majorStride = if (isTranspose) cols else rows, isTranspose = isTranspose)
   }
-  
-  def read(hc: HailContext, path: String, bufferSpec: BufferSpec): DenseMatrix[Double] = {
+
+  def read(hc: HailContext, path: String, bufferSpec: BufferSpec): BDM[Double] = {
     hc.hadoopConf.readDataFile(path)(is => read(is, bufferSpec))
   }
 
-  def apply(nRows: Int, nCols: Int, data: Array[Double], isTranspose: Boolean = false): DenseMatrix[Double] = {
-    require(data.length == nRows * nCols)
+  def readDoubles(hc: HailContext, path: String, nRows: Int, nCols: Int, rowMajor: Boolean): BDM[Double] = {
+    require(nRows * nCols.toLong <= Int.MaxValue)
+    val data = new Array[Double](nRows * nCols)
     
-    new DenseMatrix[Double](
-      rows = nRows,
-      cols = nCols,
-      data = data,
-      offset = 0,
-      majorStride = if (isTranspose) nCols else nRows,
-      isTranspose = isTranspose)
+    hc.hadoopConf.readFile(path) { is =>
+      val in = rawBufferSpec.buildInputBuffer(is)
+
+      in.readDoubles(data)
```

Comment:
You need to check the return value here.

Reply:
I'm confused. There is no return value. readDoubles will only return if `data.length` doubles have been read.